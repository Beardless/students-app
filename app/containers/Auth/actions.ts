import {
  AUTH_ACTION,
  AUTH_ACTION_ERROR,
  AUTH_ACTION_SUCCESS,
  userLoginInterface
} from "./constants";

export function authAction(
  payload: userLoginInterface
): { type: string; payload: userLoginInterface } {
  return {
    type: AUTH_ACTION,
    payload
  };
}

export function authActionSuccess(user: any): { type: string; user: any } {
  return {
    type: AUTH_ACTION_SUCCESS,
    user
  };
}

export function authActionError(
  error: string
): { type: string; error: string } {
  return {
    type: AUTH_ACTION_ERROR,
    error
  };
}
