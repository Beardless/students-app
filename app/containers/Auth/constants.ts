export const AUTH_ACTION: string = "courses/Auth/AUTH_ACTION";
export const AUTH_ACTION_SUCCESS: string = "courses/Auth/AUTH_ACTION_SUCCESS";
export const AUTH_ACTION_ERROR: string = "courses/Auth/AUTH_ACTION_ERROR";
export const USER_ACTION: string = "courses/Auth/USER_ACTION";

export interface userLoginInterface {
  username: string;
  password: string;
}
