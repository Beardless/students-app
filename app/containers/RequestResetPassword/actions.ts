import {
  REQUEST_RESET_PASSWORD,
  REQUEST_RESET_PASSWORD_ERROR,
  REQUEST_RESET_PASSWORD_SUCCESS
} from "./constants";

export function requestResetPasswordAction(
  payload: string
): { type: string; payload: string } {
  return {
    type: REQUEST_RESET_PASSWORD,
    payload
  };
}

export function requestResetPasswordActionSuccess(
  email: string
): { type: string; payload: string } {
  return {
    type: REQUEST_RESET_PASSWORD_SUCCESS,
    payload: email
  };
}

export function requestResetPasswordActionError(
  error: string
): { type: string; error: string } {
  return {
    type: REQUEST_RESET_PASSWORD_ERROR,
    error
  };
}
