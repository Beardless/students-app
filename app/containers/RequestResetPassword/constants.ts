export const REQUEST_RESET_PASSWORD: string =
  "courses/RequestResetPassword/REQUEST_RESET_PASSWORD";
export const REQUEST_RESET_PASSWORD_SUCCESS: string =
  "courses/RequestResetPassword/REQUEST_RESET_PASSWORD_SUCCESS";
export const REQUEST_RESET_PASSWORD_ERROR: string =
  "courses/RequestResetPassword/REQUEST_RESET_PASSWORD_ERROR";
