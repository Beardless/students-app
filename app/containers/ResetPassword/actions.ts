import {
  RESET_PASSWORD_ACTION,
  RESET_PASSWORD_ACTION_ERROR,
  RESET_PASSWORD_ACTION_SUCCESS
} from "./constatns";

export function resetPasswordAction(payload: {
  password: string;
  repeatedPassword: string;
  urlToken: string;
}): {
  type: string;
  payload: { password: string; repeatedPassword: string; urlToken: string };
} {
  return {
    type: RESET_PASSWORD_ACTION,
    payload
  };
}

export function resetPasswordActionSuccess(): { type: string } {
  return {
    type: RESET_PASSWORD_ACTION_SUCCESS
  };
}

export function resetPasswordActionError(
  error: string
): { type: string; error: string } {
  return {
    type: RESET_PASSWORD_ACTION_ERROR,
    error
  };
}
