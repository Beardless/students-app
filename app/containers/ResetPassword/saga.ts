import { call, put, takeEvery } from "redux-saga/effects";
import { RESET_PASSWORD_ACTION } from "./constatns";
import { getAuthorizationHeaders } from "../../utils/userUtils";
import request from "../../utils/request";
import {
  resetPasswordActionError,
  resetPasswordActionSuccess
} from "./actions";

export function* resetPassword({
  payload: { urlToken, ...rest }
}: {
  payload: { password: string; repeatedPassword: string; urlToken: string };
}) {
  try {
    const options = {
      method: "POST",
      url: `/users/password/reset?token=${urlToken}`,
      data: JSON.stringify({ ...rest }),
      headers: getAuthorizationHeaders()
    };

    yield call(request, options);

    yield put(resetPasswordActionSuccess());
  } catch (err) {
    const { message } = err.response.data;
    yield put(resetPasswordActionError(message));
  }
}

export default function* resetPasswordFlow() {
  // @ts-ignore
  yield takeEvery(RESET_PASSWORD_ACTION, resetPassword);
}
