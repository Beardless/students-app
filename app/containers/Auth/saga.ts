/**
 * Gets the repositories of the user from Github
 */

import { call, put, select, takeEvery } from "redux-saga/effects";
import { AUTH_ACTION, requestUrl, userLoginInterface } from "./constants";

// @ts-ignore
import request from "utils/request";
import { authActionError, authActionSuccess } from "./actions";
import {
  getAuthorizationHeaders,
  getToken,
  setToken
} from "../../utils/userUtils";
import { makeSelectAuthError } from "./selectors";

/**
 * Github repos request/response handler
 */
export function* getTokenAsync(payload: userLoginInterface) {
  try {
    const options = {
      method: "POST",
      body: JSON.stringify(payload),
      headers: getAuthorizationHeaders()
    };
    const res = yield call(request, `${requestUrl}/login_check`, options);

    setToken(res.token);
  } catch (err) {
    yield put(authActionError(err.message));
  }
}

export function* getUser() {
  // TODO: Add func to get user
}

export function* login({ payload }) {
  try {
    yield call(getTokenAsync, payload);

    const err = yield select(makeSelectAuthError());

    if (!err) {
      const user = { username: "Bartek" };

      yield put(authActionSuccess(user));
    }
  } catch (err) {
    yield put(authActionError(err.message));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export default function* authFlow() {
  // Watches for LOAD_REPOS actions and calls getRepos when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  // @ts-ignore
  yield takeEvery(AUTH_ACTION, login);
}
