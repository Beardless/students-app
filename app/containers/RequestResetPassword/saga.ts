import { call, put, takeEvery } from "redux-saga/effects";
import request from "../../utils/request";
import {
  requestResetPasswordActionError,
  requestResetPasswordActionSuccess
} from "./actions";
import { REQUEST_RESET_PASSWORD } from "./constants";
import { getAuthorizationHeaders } from "../../utils/userUtils";

function* requestResetPassword({ payload }: { payload: string }) {
  try {
    const options = {
      method: "POST",
      url: `/users/password/reset_request`,
      data: JSON.stringify({ email: payload }),
      headers: getAuthorizationHeaders()
    };

    console.log(options);

    yield call(request, options);

    yield put(requestResetPasswordActionSuccess(payload));
  } catch (err) {
    const { message } = err.response.data;
    yield put(requestResetPasswordActionError(message));
  }
}

export default function* requestResetPasswordFlow() {
  // @ts-ignore
  yield takeEvery(REQUEST_RESET_PASSWORD, requestResetPassword);
}
