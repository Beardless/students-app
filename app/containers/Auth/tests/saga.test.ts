import { getTokenFromApi, getUser } from "../saga";
import { userLoginInterface } from "../constants";
import { setToken } from "../../../utils/userUtils";
import { put } from "redux-saga/effects";
import { authActionError, authActionSuccess } from "../actions";
import * as user from "./mocks/user.json";

const userLoginData: userLoginInterface = {
  username: "test",
  password: "password"
};

describe("getTokenFromApi Saga", () => {
  let getTokenFromApiGenerator;

  beforeEach(() => {
    localStorage.clear();
    getTokenFromApiGenerator = getTokenFromApi(userLoginData);
    const callDescriptor = getTokenFromApiGenerator.next(userLoginData).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it("should put token in localstorage if it requests the data successfully", () => {
    const response = { data: { token: "test_token" } };
    setToken(response.data.token);
    const token = localStorage.getItem("token");
    expect(token).toEqual(response.data.token);
  });

  it("should call the authActionError if the response errors", () => {
    const err = { response: { data: { message: "Test error" } } };
    const putDescriptor = getTokenFromApiGenerator.throw(err).value;
    expect(putDescriptor).toEqual(put(authActionError("Test error")));
  });
});

describe("getUser Saga", () => {
  let getUserGenerator;

  beforeEach(() => {
    getUserGenerator = getUser();
    const callDescriptor = getUserGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it("should dispatch the authActionSuccess if it requests the data successfully", () => {
    const response = { data: user };
    const putDescriptor = getUserGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(authActionSuccess(response.data)));
  });

  it("should call the authActionError if the response errors", () => {
    const err = { response: { data: { message: "Test error" } } };
    const putDescriptor = getUserGenerator.throw(err).value;
    expect(putDescriptor).toEqual(put(authActionError("Test error")));
  });
});
