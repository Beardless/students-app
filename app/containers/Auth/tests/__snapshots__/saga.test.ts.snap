// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTokenFromApi Saga should call the authActionError if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "data": "{\\"username\\":\\"test\\",\\"password\\":\\"password\\"}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
        "url": "/login_check",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getTokenFromApi Saga should put token in localstorage if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "data": "{\\"username\\":\\"test\\",\\"password\\":\\"password\\"}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
        "url": "/login_check",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getUser Saga should call the authActionError if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "GET",
        "url": "/users/me",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getUser Saga should dispatch the authActionSuccess if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "GET",
        "url": "/users/me",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`login Saga should call the authActionError if the  1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "password": "password",
        "username": "test",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`login Saga should call the getUser saga if no errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "password": "password",
        "username": "test",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
